[Woo POS] Add POS feature flag

jaclync opened the pull request:
### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.

<!-- You can erase any parts of this template not applicable to your Pull Request. -->

### Changes proposed in this Pull Request:

<!-- If necessary, indicate if this PR is part of a bigger feature. Add a label with the format `focus: name of the feature [team:name of the team]`. -->

Replacement of WOOMOB-262 during development

Ref: pdfdoF-6VP-p2

This pull request introduces a new "Point of Sale" (POS) admin feature flag for WooCommerce mobile apps. The feature is enabled only in the development environment, and we can use this feature flag for POS features that aren't ready to be user-facing publicly. When the POS receipt feature is complete or the entire Basic POS milestone later, we can consider migrating the feature flag to a core feature switch.

<!-- Describe the changes made to this Pull Request and the reason for such changes. -->

<!-- For bug fixes: If known, please provide links to help with traceability and escape analysis. -->
<!-- Please include a link to the issue of the bug being fixed, if one doesn't exist please create it. -->
<!-- If the PR that introduced the bug is known, please also add its link below. -->


<!-- Begin testing instructions -->

### How to test the changes in this Pull Request:

<!-- Include detailed instructions on how these changes can be tested. Review and follow the guide for how to write high-quality testing instructions. -->

Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

- On the PR branch, only show one of the [core emails](https://github.com/woocommerce/woocommerce/blob/23d1d8f0ec2dda88c8e3512d26097a0d1a6b890f/plugins/woocommerce/includes/class-wc-emails.php#L241-L252) in `class-wc-emails.php` behind the admin feature flag with `if ( Features::is_enabled( 'point-of-sale' ) ) { ... }` and import `use Automattic\WooCommerce\Admin\Features\Features;` 
- Run `pnpm run --filter="@woocommerce/plugin-woocommerce" watch:build`
- Go to WooCommerce Settings > Emails (`wp-admin/admin.php?page=wc-settings&tab=email`) --> the email behind the POS feature flag should be shown
- Run `pnpm run --filter="@woocommerce/plugin-woocommerce" build`
- Go to WooCommerce Settings > Emails (`wp-admin/admin.php?page=wc-settings&tab=email`) --> the email behind the POS feature flag should **not** be shown

<!-- End testing instructions -->

### Testing that has already taken place:

<!-- Detail any testing that has already been conducted. -->
<!-- Include environment details such as hosting type, plugins, theme, store size, store age, and relevant settings. -->
<!-- Mention any analysis performed, such as assessing potential impacts on environment attributes and other plugins, performance profiling, or LLM/AI-based analysis. -->
<!-- Within the testing details you provide, please ensure that no sensitive information (such as API keys, passwords, user data, etc.) is included in this public pull request. -->

I tested the testing steps for `WC_Email_Customer_New_Account` email locally.

### Changelog entry

<!-- You can optionally choose to enter a changelog entry by checking the box below and supplying data. -->
<!-- It will trigger the 'Add changelog to PR' CI job to create and push the entry into the branch. -->

<!-- Due to org permissions, the job might fail for PRs crated from a fork under GitHub organizations. Possible solutions: -->
<!-- * Create entry manually with `pnpm --filter='@woocommerce/plugin-woocommerce' changelog add` and push it into the branch (replace `@woocommerce/plugin-woocommerce` with package name from nearest `package.json` file) -->
<!-- * Create entry from supplied PR data and push it automatically `pnpm utils changefile pr-number-here -o github-org-name-here` -->

-   [ ] Automatically create a changelog entry from the details below.

<!-- If no changelog entry is required for this PR, you can specify that below and provide a comment explaining why. This cannot be used if you selected the option to automatically create a changelog entry above. -->

-   [ ] This Pull Request does not require a changelog entry. (Comment required below)

<details>

<summary>Changelog Entry Details</summary>

#### Significance

<!-- Choose only one -->

-   [ ] Patch
-   [ ] Minor
-   [ ] Major

#### Type

<!-- Choose only one -->

-   [ ] Fix - Fixes an existing bug
-   [ ] Add - Adds functionality
-   [ ] Update - Update existing functionality
-   [ ] Dev - Development related task
-   [ ] Tweak - A minor adjustment to the codebase
-   [ ] Performance - Address performance issues
-   [ ] Enhancement - Improvement to existing functionality

#### Message <!-- Add a changelog message here -->

</details>

<details>

<summary>Changelog Entry Comment</summary>

#### Comment <!-- If your Pull Request doesn't require a changelog entry, a comment explaining why is required instead -->

</details>

jaclync committed:
Add POS feature to WC settings > Advanced > Features but not displayed in UI.

jaclync committed:
Nit cleanup changes from self review.

jaclync committed:
Merge branch 'trunk' into mobile/WOOMOB-262-add-pos-feature

# Conflicts:
#	plugins/woocommerce/src/Internal/Features/FeaturesController.php

jaclync committed:
Merge branch 'trunk' into mobile/WOOMOB-262-add-pos-feature

jaclync committed:
Remove unnecessary check of feature flag in `register` which generates a warning notice.

jaclync committed:
Fix lint errors in `PointOfSaleController`.

jaclync committed:
Replace unnecessary POS provider and controller with ja new POS feature entry.

jaclync committed:
Fix lint error in new code.

jaclync committed:
Add changelog for the POS feature flag.

kalessil committed:
Merge branch 'trunk' into mobile/WOOMOB-262-add-pos-feature

gigitux committed:
Blocks: fix integration test framework (#57458)

* feat: add block setting to control spec visibility

* chore: changelog

* chore: phpcs

* test: verify toggle behaviors

* Update plugins/woocommerce/changelog/wooplug-3844-add-conditional-options-for-displaying-weight-dimensions-and

Co-authored-by: Karol Manijak <20098064+kmanijak@users.noreply.github.com>

* chore: remove redundant fontSize attribute

* Blocks - integration tests: ensure that Woo blocks aren't unregistered

* improve integration test framework

* improve unit test

* Add changefile(s) from automation for the following project(s): woocommerce

* fix conflict

* lint code

---------

Co-authored-by: Tung Du <dinhtungdu@gmail.com>
Co-authored-by: Karol Manijak <20098064+kmanijak@users.noreply.github.com>
Co-authored-by: github-actions <github-actions@github.com>

chihsuan committed:
[Blueprint] Implement event tracking and fix error reporting (#57336)

* Enhance Blueprint error handling and tracking in upload and export processes

* Update error messages to use translation functions for better localization.
* Implement event tracking for successful and failed blueprint imports and exports.
* Refactor error reporting to provide more detailed feedback on import/export issues.
* Add translation support for various error messages related to blueprint file handling.

* Add changefile(s) from automation for the following project(s): woocommerce, woocommerce/client/admin

* Refactor error handling in Blueprint settings and update styles

* Rename error state variable for clarity from `error` to `exportError`.
* Update error handling to display export errors with a new notice component.
* Remove unused error notice from the upload dropzone.
* Enhance styling for error notices in settings to improve visibility and consistency.

---------

Co-authored-by: github-actions <github-actions@github.com>
Co-authored-by: Vladimir Reznichenko <kalessil@gmail.com>

AhmarZaidi committed:
Fix password-protected product layout in blockified templates shows unnecessary product title and product excerpt (#57437)

* fix: remove password protected product title and excerpt

* add: changelog entry

* add: check to remove empty inner blocks

* update: hide logic for related products pattern

* update: test for single product template

---------

Co-authored-by: Vladimir Reznichenko <kalessil@gmail.com>

chihsuan committed:
[Blueprint] Fix php warnings and improve error handling (#57268)

* Enhance UseWPFunctions trait with additional filesystem methods and improved documentation

- Added `wp_filesystem_get_contents` method for reading file contents.
- Updated return types in existing methods for clarity.
- Improved documentation comments for better understanding of method behaviors.

* Refactor ExportCli to utilize UseWPFunctions for filesystem operations

- Integrated UseWPFunctions trait into ExportCli class.
- Replaced direct file_put_contents call with wp_filesystem_put_contents for improved filesystem handling.
- Enhanced error handling for file saving operations.

* Refactor ResourceStorages class to improve parameter naming and enhance clarity

- Updated method parameters from `$resource` to `$resource_type` for better understanding.
- Adjusted logic in `is_supported_resource` and `download` methods to reflect the new parameter naming.
- Improved code readability by removing unnecessary comments and ensuring consistent formatting.

* Enhance OrgPluginResourceStorage class with improved error handling in download methods

- Updated return types in the `download` and `download_url` methods for clarity, changing from nullable types to more explicit error handling.
- Added checks for `WP_Error` to ensure robust error management during plugin downloads.
- Improved documentation comments to reflect the changes in return types and enhance understanding of method behaviors.

* Improve error handling in OrgThemeResourceStorage class by checking for WP_Error before accessing download link

- Added a check for WP_Error to return null if an error occurs, enhancing the robustness of the download link retrieval process.
- This change ensures that the method handles errors gracefully, preventing potential issues when accessing the download link.

* Update documentation for plugin activation and deletion methods in UsePluginHelpers trait

- Revised return type annotations for `activate_plugin_by_slug` and `delete_plugin_by_slug` methods to provide clearer expectations.
- Updated comments to reflect the possibility of returning `WP_Error` and `null`, enhancing understanding of error handling in these methods.

* Refactor plugin and theme activation processes for improved clarity

- Updated comments in `ImportActivatePlugin` to clarify the use of non-snake case property names.
- Modified `ImportActivateTheme` to remove the unnecessary return value from `wp_switch_theme`, enhancing code readability and understanding of the method's behavior.

* Refactor theme installation method for improved clarity

- Updated method parameter name from `$local_plugin_path` to `$local_path` to accurately reflect its purpose.
- Revised documentation comments to clarify that the method installs a theme from the local path, enhancing understanding of its functionality.

* Improve error handling in ImportDeactivatePlugin by checking for WP_Error after deactivating a plugin

- Added a check for WP_Error when attempting to deactivate a plugin, providing clearer feedback on success or failure.
- Updated the response messages to inform users whether the deactivation was successful or if an error occurred.

* Enhance JsonSchema class with improved error handling and filesystem integration

- Integrated UseWPFunctions trait to utilize WordPress filesystem functions.
- Added checks for valid schema path and improved error handling for JSON file reading.
- Updated validation to ensure 'steps' field is an array, enhancing schema integrity.

* Add changelog

* Refactor theme activation process to improve clarity and remove unnecessary return value

- Removed the return value from the `wp_switch_theme` method in `ImportActivateTheme`, enhancing code readability.
- Added a check to confirm the theme switch was successful before logging the debug message, improving the accuracy of feedback provided to users.

* Add tests for JsonSchema class to improve error handling coverage

- Introduced new test methods to validate exception handling for invalid paths and unreadable files.
- Utilized Mockery to simulate filesystem behavior, enhancing the robustness of the tests.
- Ensured that the JsonSchema class correctly throws exceptions for various error scenarios, improving overall error handling.

* Enhance ImportInstallPlugin with improved error handling and clarity

- Added checks to skip installation if the plugin is already installed and to validate resource types.
- Improved error handling during plugin installation and activation, providing clearer feedback on success or failure.
- Updated method documentation to reflect the possibility of returning WP_Error on failure.

* Implement error handling in ImportCli for schema file creation

- Added try-catch block in the run method to handle exceptions when creating the blueprint from the schema file.
- Improved user feedback by utilizing WP_CLI::error to display error messages, enhancing the robustness of the import process.

---------

Co-authored-by: Vladimir Reznichenko <kalessil@gmail.com>

scruffian committed:
Add to cart variations: Don't break word (#57274)

* Add to cart variations: Don't break word

* move CSS to the blocktheme file

* add changelog

elazzabi committed:
WOOPLUG-3201: Add GrabPay to WooPayments supported payment methods (#57436)

* Add SVG icons for GrabPay and WeChat Pay payment methods

* Add GrabPay and WeChat Pay payment methods to WooPaymentsMethodsLogos component

* Add changefile(s) from automation for the following project(s): @woocommerce/onboarding

* Clean GrabPay SVG

* remove className from SVG path

* Revret the addition of WeChat Pay

* Add changefile(s) from automation for the following project(s): @woocommerce/onboarding, woocommerce

* Delete the correct wechat-pay file

* Update the total PMs count

* Add changefile(s) from automation for the following project(s): @woocommerce/onboarding

---------

Co-authored-by: github-actions <github-actions@github.com>
Co-authored-by: Vladimir Reznichenko <kalessil@gmail.com>

mikejolley committed:
[Checkout] Improve Delivery selector styling and spacing (#56789)

* Update colors in prep for using in shipping method selector

* Updated styling

* Changelog

* Improve wrapping

* Use input text colors for toggle

* Prevent scroll on shipping/pickup toggle

* Update naming

* Change text wrapping for toggle titles

tjcafferkey committed:
ShippingController->register_local_pickup defensive checks to prevent fatal (#57154)

* Defensive check against method dependencies

* Add changelog

* use method_exists

* Correctly case WC()

* Log error and test for this

* test setup to ensure block checkout is in use

* Doc block for tear down

* Move test setup into setUp() and tearDown() methods

* Check for WC

* is_object check

* Check if shipping is an object

csmcneill committed:
Update Extensibility in Blocks documentation to resolve 404 errors (#57492)

* Update Extensibility in Blocks documentation to resolve 404 errors

Originally reported on Slack at p1745429460671199-slack-C8X6Q7XQU

Many of the links on this page use a relative path (e.g., `../../plugins/woocommerce/client/blocks/`) to link out to Markdown files hosted in the GitHub repository.

When the page is published, the links use the relative path of `developer.woocommerce.com`, but since these pages don't exist, the resulting pages can't be found, generating a 404 error.

In this change, the links are updated to remove the relative path structure and they include the full GitHub Markdown file URLs. This will prevent 404 errors from occurring when navigating through this document.

* Update docs manifest

---------

Co-authored-by: Vladimir Reznichenko <kalessil@gmail.com>

nerrad committed:
Update project-community-pr-assigner.yml - update to new team labels.

chihsuan committed:
[Blueprint] Improve schema/filter validation  (#57270)

* Enhance ExportSchema class with landing page validation and exporter filtering

- Added validation for the landing page path to ensure it is a valid relative local URL, throwing an InvalidArgumentException for invalid paths.
- Refactored the landing page assignment to use a variable for clarity.
- Implemented filtering of exporters to ensure only instances of StepExporter are included, improving the integrity of the export process.

* Enhance ImportSchema class with improved error handling and validation

- Added exception documentation for the `create_from_file` and `create_from_json` methods to clarify potential errors.
- Implemented validation for step processors to ensure they are instances of StepProcessor, improving the integrity of the import process.
- Updated JSON validation to use `wp_json_encode`, enhancing compatibility with WordPress standards.

* Validate importer type in ImportStep to ensure it is a valid StepProcessor before processing, enhancing error handling and integrity of the import process.

* Enhance ExportSchema and ImportStep tests with additional validation scenarios

- Updated ExportSchemaTest to validate landing page path and throw an exception for invalid paths.
- Added a new test in ImportStepTest to ensure warnings are returned when an importer is not a valid StepProcessor, improving test coverage and error handling.

* Refactor ExportSchema to return WP_Error for invalid landing page paths

- Updated the export method to return a WP_Error instead of throwing an exception for invalid landing page paths, improving error handling.
- Modified the corresponding test to validate the new behavior, ensuring it returns a WP_Error instance when the path is invalid.

* Improve error handling in ExportCli by checking for WP_Error in export result

- Updated the export method in ExportCli to handle WP_Error responses from ExportSchema, providing clearer error messages in the CLI.
- Ensured that the export result is properly encoded and saved only if no errors occur, enhancing the robustness of the export process.

* Add test to filter out invalid exporters in ExportSchemaTest

- Implemented a new test method to ensure that only valid instances of StepExporter are included in the export process.
- Added mock exporters for testing, including an invalid exporter that throws an exception, to validate the filtering logic.
- Enhanced test coverage for the ExportSchema class by verifying the correct handling of invalid exporters.

* Add changelog

dinhtungdu committed:
Product Filters Checkbox List: review the usage of `:where` usage (#57497)

Using :where selector for form element make the checkbox list style
easier to break. We reduced it to only custom class for generic elements
like div and span to strike a balance, keeping our style consistence and
still allowing easy overriding.

gigitux committed:
Hide 'View Cart' button when the cart page is not defined (#57495)

* Hide 'View Cart' button when the cart page is not defined

* Add changefile(s) from automation for the following project(s): woocommerce

* add unit test

* fix unit test

---------

Co-authored-by: github-actions <github-actions@github.com>

chihsuan committed:
[Blueprint] Add permission validation for import steps (#57294)

* Refactor ImportSchema and ImportStep classes to enhance permissions validation

- Removed the built-in step processors from ImportSchema, simplifying the class structure.
- Introduced a new ImportStep class to handle step processing, improving modularity.
- Added capability checks in step processors to ensure users have the necessary permissions before executing steps.
- Updated validation methods to return boolean values, enhancing clarity on validation outcomes.

* Update ImportStepTest to extend WP_UnitTestCase and add capability validation test

- Changed ImportStepTest to extend WP_UnitTestCase for better integration with WordPress testing framework.
- Added a new test to validate error handling when user capabilities are insufficient for executing a step, enhancing test coverage and robustness of the import process.
- Introduced a check_step_capabilities method in DummyImporter to support capability validation.

* Fix function existence check in CliResultFormatter to use namespaced format_items function

- Updated the function existence check for format_items to use the correct namespaced path '\WP_CLI\Utils\format_items'.
- Reorganized the conditional logic to throw an exception if the function is not found before calling it, improving error handling and clarity.

* Update ImportStep to use step definition in success result

- Modified the import method to utilize the step definition's step property in the success result, enhancing clarity and ensuring the correct step context is used during the import process.

* Add changefile(s) from automation for the following project(s): packages/php/blueprint

* Refactor capability checks in ImportRunSql to check all permissions

* Refactor capability check formatting in ImportDeletePlugin for improved readability

* [Blueprint] Add permission validation for export steps and REST API (#57303)

* Enhance ExportSchema and related exporters with capability checks and error handling

- Updated the export method in ExportSchema to return WP_Error for invalid landing page paths instead of throwing exceptions, improving error handling.
- Added check_step_capabilities method to various StepExporter implementations to validate user permissions before executing steps, ensuring only authorized users can perform exports.
- Updated CLI export command to handle WP_Error responses gracefully, providing clear error messages to the user.
- Enhanced documentation for methods to clarify expected return types and capabilities checks.

* Refactor permission checks in RestApi for export and import actions

- Updated permission callbacks for export and import methods to use specific checks: `check_export_permission` and `check_import_permission`.
- Enhanced permission validation to ensure only users with appropriate capabilities can perform export and import actions, improving security and clarity in user permissions.

* Improve error handling in ExportSchema and RestApi

- Updated error message in ExportSchema to clarify the context of insufficient permissions during export steps.
- Added handling for WP_Error responses in RestApi to return appropriate HTTP responses, enhancing API robustness and user feedback.

* Improve error handling in slotfill.js for export functionality

- Enhanced error handling in the Blueprint component to differentiate between various error types, including string errors and specific permission-related errors.
- Added user-friendly error messages for insufficient permissions during export, improving clarity for users.
- Ensured that the export functionality is re-enabled after error handling, maintaining a smooth user experience.

* Enhance steps_payload_to_blueprint_steps method in RestApi

- Added checks to ensure that 'settings', 'plugins', and 'themes' arrays are not only set but also contain elements before processing them, improving the robustness of the steps payload handling.

* Refactor error handling in ExportSchema for consistency

- Updated the ExportSchema class to use the WP_Error class consistently without the namespace prefix, improving code readability and maintaining consistency across error handling.
- Clarified return type documentation for the export method to reflect the use of WP_Error, enhancing developer understanding of the method's behavior.

* Initialize error state in exportBlueprint function in slotfill.js

- Added a line to reset the error state to null before starting the export process, ensuring that previous errors do not affect the current export operation.

* Add changelog

* Fix tests

* Enhance DummyExporter with strict types and additional methods

- Added strict types declaration for improved type safety.
- Introduced check_step_capabilities method to validate export capabilities.
- Enhanced documentation for existing methods to clarify their purpose and return types, improving code readability and developer understanding.

* Fix lint

* [Blueprint] Implement logging for export and import operations (#57384)

* Implement logging functionality in ExportSchema and ImportStep classes

- Introduced a Logger class to handle logging of export and import steps, improving traceability and debugging.
- Enhanced ExportSchema to log the start and completion of export steps, as well as errors encountered during the process.
- Updated ImportStep to log the start and result of the import process, including success and error messages.
- Refactored error handling in both classes to utilize the new logging system, ensuring consistent logging practices across the codebase.

* Add changelog

* Fix tests

* Refactor Logger class to utilize WP functions

- Introduced UseWPFunctions trait to encapsulate WordPress function usage.
- Updated Logger class to use wp_get_current_user_id() for retrieving the current user's ID, enhancing consistency with WordPress standards.
- Minor adjustment in ImportStepTest class to ensure proper namespace usage for WP_UnitTestCase.

* Refactor logging in ExportSchema and ImportStep classes

- Updated ExportSchema and ImportStep to utilize new logging methods in the Logger class for starting, completing, and handling failures of export and import steps.
- Removed redundant logging code and centralized logging functionality in the Logger class, enhancing code clarity and maintainability.

* Remove WC_Log_Levels and WC_Logger_Interface stubs; centralize logging in tests

- Deleted the WC_Log_Levels and WC_Logger_Interface stub files to streamline the codebase.
- Introduced a new stubs.php file to provide necessary stubs for testing, ensuring compatibility and reducing redundancy.
- Updated references in the Logger class to use fully qualified class names for WC_Log_Levels, enhancing clarity and consistency.

* Fix tests

* Add changefile(s) from automation for the following project(s): packages/php/blueprint, woocommerce, woocommerce/client/admin

---------

Co-authored-by: github-actions <github-actions@github.com>

amitraj2203 committed:
[Experimental] Blockified Update Quantity Selector and Grouped Product Selector to use Stepper mode only (#57442)

* refactor: remove input mode from quantity selector, allowing only stepper mode

* feat: Add stepper mode and remove input mode

* feat: Add changelog message

* refactor: Quantity input handling in Add to Cart blocks to utilize utility methods for improved maintainability and readability

* refactor: Consolidate quantity handling logic and remove redundant frontend file

* refactor: Remove redundant quantity selector styles and consolidate into main style file

* refactor: Introduce QuantitySelector component to consolidate the component between different blocks

* refactor: Remove unused style import from quantity selector component

* refactor: Update description in quantity selector block to clarify input functionality

* refactor: Simplify type assertion for input element in getInputElementFromEvent function

* refactor: Rename quantity action methods for consistency and clarity in AddtoCartWithOptions block

* refactor: Consolidate quantity selector styles and remove unused files

* refactor: Wrap all the styles in this file with the two blocks parent classes

* refactor: Simplify quantity selector style assignment in downgrade notice

* refactor: Remove unused script enqueue method from grouped product selector CTA

* fix: Add check for $previous_product is an instance of WC_Product.

* refactor: Remove isSiteEditor prop from QuantityStepper component

* refactor: Simplify product retrieval in context methods and update validation logic

* fix: Correct namespace casing for AddtoCartWithOptions in relevant files

* fix: Linitng error

---------

Co-authored-by: Vladimir Reznichenko <kalessil@gmail.com>

chihsuan committed:
[Blueprint] Add restrictions for options in Blueprint importer (#57381)

* Implement restricted options handling in ImportSetSiteOptions

- Added a list of restricted WordPress options that cannot be modified for security reasons.
- Updated the process method to skip updates for these restricted options and log warnings accordingly.
- Enhanced unit tests to verify that restricted options are not modified and appropriate warnings are generated.

* Add changelog

* Change message

dinhtungdu committed:
Product Details: only render non-empty accordion items (#57516)

chihsuan committed:
[Blueprint] Enhance SQL execution safety (#57344)

* Enhance SQL execution safety in ImportRunSql class

* Introduced validation checks for allowed SQL query types.
* Added mechanisms to detect SQL injection patterns and prevent unauthorized modifications to protected user tables and capabilities.
* Implemented SQL normalization to standardize query format.
* Enhanced error handling during SQL execution with transaction support.

* Add unit tests for ImportRunSql class

* Implemented comprehensive tests for SQL query processing, including valid INSERT, UPDATE, and REPLACE queries.
* Added tests to ensure invalid query types are correctly rejected and appropriate error messages are returned.
* Included tests for SQL normalization, handling of SQL injection patterns, and protection against modifications to user roles and capabilities.
* Ensured proper error handling for SQL execution failures and multiple statement queries.

* Add changelog

* Improve error handling in ImportRunSql class during SQL execution

* Updated error suppression to be enabled explicitly.
* Refactored error handling to store the last error in a variable for clarity.
* Enhanced error messages to provide more informative feedback on SQL execution failures.

* Refactor SQL handling in ImportRunSql class for improved security and clarity

* Removed the normalize_sql method and replaced it with direct SQL trimming.
* Added checks for suspicious SQL comments to enhance security against hidden malicious code.
* Updated related tests to validate detection of suspicious comments and ensure proper error handling for SQL queries.
* Simplified SQL validation logic by directly using trimmed SQL content in various checks.

* Fix tests

* Refactor SQL query preparation in ExportWCTaxRates class for improved security

* Updated the SQL query in generateSteps method to use prepared statements, enhancing security against SQL injection.
* Removed direct table name interpolation to ensure safer query execution.

* Update table name handling in ExportWCTaxRates class for improved SQL query preparation

* Modified the generateSteps method to prepend the table prefix directly to the table name variable, ensuring consistent and secure SQL query execution.
* This change enhances clarity and maintains the integrity of the SQL preparation process.

* Reformat

* Add changelog

* Remove unused imports

* [Blueprint]  Add security restriction to only allow blueprint imports in Coming Soon mode (#57382)

* Implement import permission check for blueprint uploads

- Added a new API endpoint to check if blueprint imports are allowed based on site status.
- Integrated the import permission check into the file upload state machine.
- Updated the UI to display a notice when imports are disabled, guiding users to enable the "Coming Soon" mode or set a constant for live sites.
- Enhanced styling for error and notice components related to blueprint uploads.

This change improves user experience by providing clear feedback on import permissions.

* Add changelog

* Implement import permission check in the RestApi class

- Added a check in the import_step method to verify if blueprint imports are allowed.
- Returns an error message if imports are disabled, enhancing user feedback during the import process.

* Fix lints

* Add unit tests for blueprint import functionality

- Introduced tests to validate the import_step endpoint, including checks for file size limits and import permissions based on site status.
- Enhanced the setup and teardown methods for better resource management during tests.
- Ensured that appropriate error messages are returned when imports are disabled or when file size exceeds the limit, improving feedback for users.

* Fix lint

jorgeatorres committed:
Bump stable tag to 9.8.2 (#57524)

pavel-mailpoet committed:
Add Woo email content to email preview (#57337)

* Replace placeholder content in email previews

Ensure email previews display dynamic content using actual email generation logic.
Improves preview accuracy and adds nonce verification for security.

[WOOPLUG-3316]

* Get the email type from post when sending preview

[WOOPLUG-3316]

* Refactor email preview content generation

Consolidate the logic for generating email placeholder content
used in different parts of the email editor and preview features.
This avoids code duplication and centralizes the generation process.

[WOOPLUG-3316]

* Add changefile(s) from automation for the following project(s): packages/php/email-editor, woocommerce

* Initialise a variable before usage

[WOOPLUG-3316]

* Get email type form post meta

[WOOPLUG-3316]

* Document the preview filter

[WOOPLUG-3316]

* Add changefile(s) from automation for the following project(s): packages/php/email-editor, woocommerce

* Improve changelog

[WOOPLUG-2216]

* Add changefile(s) from automation for the following project(s): packages/php/email-editor, woocommerce

---------

Co-authored-by: github-actions <github-actions@github.com>
Co-authored-by: Vladimir Reznichenko <kalessil@gmail.com>

amitraj2203 committed:
[Experimental] Blockified Add to Cart with Options block: Add utility method to check if a simple product is not purchasable (#57523)

* refactor: AddToCartWithOptions logic to use utility method for checking non-purchasable simple products

* feat: Rename AddtoCartWithOptions folder name

* feat: Rename the files where Utils class of AddToCartWithOptions is being used.

* feat: Add changelog file

joshuatf reviewed (commented):
Hey @jaclync 👋  The code in here looks good and happy to approve as is if this is temporary, though I'm curious if we plan on adding a feature flag (under the feature flag controller) in for POS.

My concern is that if someone begins using this as the source of truth for POS and this is used to determine if other POS UI is shown, we may have functionality bleeding through on non-mobile based apps.

If we do introduce that flag, it might be good to conditionally enable this based on that feature availability.

dmallory42 committed:
Add new Bank Details Table Component (#57379)

* WIP

* WIP

* WIP

* Updates to support bank account list component.

* Updates to support bank account list component.

* Removing code used for testing.

* Removing code used for testing.

* Update changelog.

* Add newline.

* Styling fixes for the bank account modal.

* Form validation.

* Form validation.

* Form validation.

* Form validation.

* Remove accidentally committed code and lint fix

* One more lint fix

* Update to add some JSDocs.

---------

Co-authored-by: Vladimir Reznichenko <kalessil@gmail.com>
Co-authored-by: Oleksandr Aratovskyi <79862886+oaratovskyi@users.noreply.github.com>
Co-authored-by: oaratovskyi <oleksandr.aratovskyi@automattic.com>

chihsuan committed:
Fix failed opening required `blueprint/tests/stubs/stubs.php` (#57527)

* Skip tests

* Add changelog

* Fix changelog

pavel-mailpoet committed:
Hide backorder meta on completed orders (#57450)

* Add changelog

[WOOPLUG-266]

* Hide backorder meta on completed orders

Products previously on backorder continued showing the backorder status
in emails after the order was fulfilled.

[WOOPLUG-266]

* Use an enum instead of hardcoded string

[WOOPLUG-266]

* Use an enum instead of hardcoded string in test

[WOOPLUG-266]

* Remove redundant code

[WOOPLUG-266]

---------

Co-authored-by: Vladimir Reznichenko <kalessil@gmail.com>

kmanijak committed:
Product Gallery: Add alignment and layout options to Next/Previous buttons (#57484)

* Add support for alignment, layout and margin

* Adjust styles

* Handle alignment in editor

* Adjust frontend to changes

* Change the spread of properties so actual buttons are the block and container is just used for alignment

* Provide alignment to container on frontend

* Improve type and add some context

* Add changelog

* Remove local radius style so theme has more freedom

* Add experimental selector for global styles

* Add z-index for editor pseudo-elements to work correctly

* Pass blockProps to container but provide proper classNames and styles to buttons

* Refactor class split into more robust and future proof version filtering out local styles classes

* Add comment with some context

* Fix outline that may be cut off when buttons are too close

* Exclude alignment styles from buttons

* Skip serialization of supports properties so we don't need to filter the properties out from container classnames

* Fix lint

* Remove unnecessary z-index preventing from image interactions

* Fix Next/Prev buttons selector

---------

Co-authored-by: Vladimir Reznichenko <kalessil@gmail.com>

jorgeatorres committed:
Add 'id' attribute to settings of type 'single_select_country' (#57247)

* Make sure fields of type 'single_select_country' have an 'id' attribute.
* Add changelog

jaclync committed:
Replace core feature switch with admin feature flag that can only be enabled only in development environment.

jaclync committed:
Merge branch 'trunk' into mobile/WOOMOB-262-add-pos-feature

jaclync committed:
Update changelog.

jaclync commented:
❓  One question about the admin feature flags, is there a REST API for admin/shop manager users by any chance? From my search I didn't find one but wanted to ask in case there's a way.

joshuatf reviewed (approved):
LGTM!   Tested with the following snippet:

```php
add_action( 'admin_init', function() {
	if ( \Automattic\WooCommerce\Admin\Features\Features::is_enabled( 'point-of-sale' ) ) {
		error_log( 'point-of-sale is enabled' );
	}
} );
```
And toggled the flags on/off in the config file.  The log was shown when the flag was enabled and not shown when disabled as expected.

joshuatf commented:
> ❓ One question about the admin feature flags, is there a REST API for admin/shop manager users by any chance? From my search I didn't find one but wanted to ask in case there's a way.

It's not exactly RESTful, but there is an API for adding features to the UI if you want to allow merchants to toggle the feature on/off [here](https://github.com/woocommerce/woocommerce/blob/6eea6051d3162033ea8a7dbdea6dd2bb71bc2b69/plugins/woocommerce/src/Internal/Features/FeaturesController.php/#L240-L464).  Is that what you're looking for?

jaclync commented:
> Tested with the following snippet:

Thanks for sharing a simpler way to test! Where do you usually put such a snippet when testing a PR?

> It's not exactly RESTful, but there is an API for adding features to the UI if you want to allow merchants to toggle the feature on/off [here](https://github.com/woocommerce/woocommerce/blob/6eea6051d3162033ea8a7dbdea6dd2bb71bc2b69/plugins/woocommerce/src/Internal/Features/FeaturesController.php/#L240-L464). Is that what you're looking for?

I see, yea I believe we will eventually add this feature switch for POS when the work is complete as we'd like to enable it by default.

---

Merging the PR now and the next one will include the feature flag usage.

joshuatf commented:
I think most people use a plugin like [this one](https://wordpress.org/plugins/code-snippets/), but any workflow will do.  Custom plugin, must use plugin, or honestly in any code that gets run within plugin context.
